### PR TITLE
Rework relationship between citrine-python & gemd-python's serialization layers

### PR DESCRIPTION
**Pull request by kroenlein**
# Citrine Python PR

## Description 
This PR creates a cleaner surface separating gemd-python & citrine-python's serialization behaviors.

* It updates the GEMD-derived objects to use the implicit typ system introduced in gemd-python 1.13.0
* It makes the MRO consistently choose serialization, then citrine-python objects, then gemd-objects
* It updates the _serialization tools to respect mro (method resolution order) and to search anywhere in the class structure for a method instead of just one layer deep.
* It adds an `use_init` boolean to Property objects to flag those fields that require an `__init__` routine to be run for proper initialization.
* It corrects a number of what looked like previously uncovered bugs.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Maintenance (non-breaking change to assist developers)

### Adherence to team decisions
- [X] I have added tests for 100% coverage
- [X] I have written Numpy-style docstrings for every method and class.
- [X]  I have communicated the downstream consequences of the PR to others.
- [X] I have bumped the version in __version\__.py

### CODE REVIEW

File: src/citrine/_rest/resource.py
Comment (kroenlein) [inline src/citrine/_rest/resource.py:54]: All GEMD objects / DictSerializables go through this one Resource to maintain a consistent interface layer.

File: src/citrine/_rest/resource.py
Comment (kroenlein) [inline src/citrine/_rest/resource.py:72]: Support Material History-like serializations.  Handles both the return structure from the `material_history` route as well as the GEMD denormalized format.

File: src/citrine/_rest/resource.py
Comment (kroenlein) [inline src/citrine/_rest/resource.py:76]: This is new behavior, and I'm open to dropping it.  It essentially says that `ProcessSpec.build` expects that you're actually passing it a `ProcessSpec`, as opposed to e.g. a `MaterialRun`.  The alternative is to propagate Polymorphic behavior for build to all concrete GEMD objects.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:56]: New parameter.  When de-serializing an object, use klass(args) instead of __new__.  Necessary for cases where, for example, a private field is set to None in the __init__ routine.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:62]: With the properties clean up, `override` became redundant with the None-ish value of `_key`.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:137]: This was a latent bug.  `serialization_path` happened to equal all property_names for the cases where something was overridden, but that was definitely not guaranteed.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:232]: This caching drops test execution share from 28 % to 6 %.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:234]: Previously, we were incorrectly searching for serialization_path.  This first finds the actual field key and then resolves the MRO base class (instead of demanding that the property exist along exactly one inheritance path).

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:696]: Uses a reverse MRO instead of piecemeal pulling things in from two most immediate layers.  Allows for proper method inheritance instead of just a shallow inheritance.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:689]: Caching drops this routine's testing share from 30 % to 7.5 %.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:838]: Modify LinkOrElse to actually keep track of what Else means.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [inline src/citrine/_serialization/properties.py:857]: No need for LinkOrElse to know how to decode a LinkByUID.

File: src/citrine/resources/api_error.py
Comment (kroenlein) [inline src/citrine/resources/api_error.py:7]: DictSerializable is being modified to make objects w/o typ abstract according to ABC.  There is no type field in the object that comes in.  Might be reasonable to just migrate this to a Serializable[] object.

File: src/citrine/resources/api_error.py
Comment (kroenlein) [inline src/citrine/resources/api_error.py:18]: DictSerializable is being modified to make objects w/o typ abstract according to ABC.  There is no type field in the object that comes in.  Might be reasonable to just migrate this to a Serializable[] object.

File: tests/_serialization/_data.py
Comment (kroenlein) [inline tests/_serialization/_data.py:89]: Modified to be valid, since deserialization routines aren't bothered by previously deserialized values anymore.

File: tests/_serialization/test_container_properties.py
Comment (kroenlein) [inline tests/_serialization/test_container_properties.py:73]: serialization_path changed to required argument for top-level properties.  Doesn't make sense to have a top-level property without it.

File: src/citrine/resources/data_concepts.py
Comment (kroenlein) [inline src/citrine/resources/data_concepts.py:41]: We have to do weird things here because GEMD.typ is a class-value, rather than an object value, property, or classmethod.  Ideally, it'd be a classmethod+property, but Python 3.12 dropped support for that concept.

File: src/citrine/resources/data_concepts.py
Comment (kroenlein) [inline src/citrine/resources/data_concepts.py:38]: Once we can import a gemd-python version that has abstract + concrete classes, this can be simplified.  That is expected to be released next cycle.

File: src/citrine/_rest/resource.py
Comment (bfolie) [inline src/citrine/_rest/resource.py:60]: Which situations does this apply to?

File: src/citrine/_serialization/properties.py
Comment (bfolie) [inline src/citrine/_serialization/properties.py:42]: Missing docstring for `default`

File: src/citrine/_serialization/properties.py
Comment (bfolie) [inline src/citrine/_serialization/properties.py:103]: In what situation are we trying to deserialize an already-deserialized object?

File: src/citrine/_rest/resource.py
Comment (kroenlein) [reply src/citrine/_rest/resource.py:60]: https://github.com/CitrineInformatics/gemd-python/blob/b625aee87d02768d918ccb4088a97b43fa15c894/gemd/json/gemd_json.py#L32
gemd-python makes a material history with a `object` key for the root object of the serialized graph
https://github.com/CitrineInformatics/citrine-python/blob/9be3cded48ea64f754d1897513100fd2230d0c52/src/citrine/resources/material_run.py#L141
platform returns a material history with a `root` key for a LinkByUID to the root object.

File: src/citrine/resources/api_error.py
Comment (bfolie) [reply src/citrine/resources/api_error.py:7]: It's a smell to have this citrine-python Error object inheriting its type from GEMD

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:103]: The `flatten` utility from gemd-python only partly serializes its objects when dissecting them, the validate_templates method in citrine-python does something similar, and many tests use build combined with Objects in the incoming values stream.

Maybe not ideal, but correcting this would make this PR substantially larger.  Not a new behavior.

File: src/citrine/resources/api_error.py
Comment (kroenlein) [reply src/citrine/resources/api_error.py:7]: Probably better to rebuild from scratch.  This seemed like the highest-velocity solution.

File: src/citrine/_serialization/properties.py
Comment (bfolie) [inline src/citrine/_serialization/properties.py:111]: Why is resolving the base class no longer necessary along this code path?

File: src/citrine/_serialization/properties.py
Comment (bfolie) [reply src/citrine/_serialization/properties.py:232]: Why 1024*1024?

File: src/citrine/_serialization/properties.py
Comment (bfolie) [inline src/citrine/_serialization/properties.py:755]: Is there a way to make this based on the specific error class, as opposed to string-matching on the message?

File: src/citrine/_serialization/properties.py
Comment (bfolie) [reply src/citrine/_serialization/properties.py:857]: Why not?

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:111]: https://github.com/CitrineInformatics/citrine-python/blob/9be3cded48ea64f754d1897513100fd2230d0c52/src/citrine/_serialization/properties.py#L666-L686

There's no way to get to that line where `data` is not a `dict`.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:755]: Error condition is that you've failed to pass a required argument to __init__.  This fires off `TypeError` with messges of the form 
`TypeError: CLASS.__init__() missing 1 required positional argument: 'arg_name'`
or 
`TypeError: CLASS.__init__() missing 1 required keyword-only argument: 'arg_name'`
I thought just catching a TypeError would catch too many other potential issues.  There's no tighter error type to filter by.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:232]: Arbitrarily large while not very large.  A tighter estimate would be total number of property objects in all the files, but I figured trying to guess at that wasn't worth the effort.

File: src/citrine/_serialization/properties.py
Comment (bfolie) [reply src/citrine/_serialization/properties.py:755]: What I'm wondering is if we could change those from being TypeErrors to being some other type and then recognize that?

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:857]: Because it already needs to know to ask `DictSerializable` to decode the object, and `DictSerializable` knows how to decode LinkByUIDs.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:755]: We'd need to scan the __init__ argument list and compare what we're about to pass it.  This is actually not terrible because python provides [https://docs.python.org/3.10/library/inspect.html#inspect.getfullargspec] and [https://docs.python.org/3.10/library/inspect.html#inspect.signature], but I, again, trying to maintain a bit of velocity.

Comment (kroenlein) [reply src/citrine/_serialization/properties.py]: Test implemented with [signature](https://docs.python.org/3.10/library/inspect.html#inspect.signature).

File: src/citrine/resources/api_error.py
Comment (kroenlein) [reply src/citrine/resources/api_error.py]: Classes rebuilt as per _serialization.

File: tests/resources/test_gemd_resource.py
Comment (bfolie) [inline tests/resources/test_gemd_resource.py:149]: This is also checking that register appends a uid to the returned object, but not the original object, yes?

Comment (kroenlein) [reply tests/resources/test_gemd_resource.py:149]: I believe the intended behavior in GEMD object registration is to edit the incoming objects such that they also contain new identifiers.  For objects with no UIDs, the act of serialization (all client-side) requires that they have at least one UID and CitrineIDs are assigned client side.  That alone would modify the objects.

This may be inconsistent with other behaviors in the library, and is certainly inconsistent with a `no side effects` design paradigm, but gemd-python doesn't follow a `no side effects` policy around serialization.

File: tests/serialization/test_material_run.py
Comment (bfolie) [inline tests/serialization/test_material_run.py:191]: outdated debugging?

File: tests/serialization/test_material_run.py
Comment (bfolie) [inline tests/serialization/test_material_run.py:155]: It looks like this docstring is no longer valid

File: tests/_serialization/test_object_serialization.py
Comment (bfolie) [inline tests/_serialization/test_object_serialization.py:81]: What's going on here?

File: src/citrine/_serialization/properties.py
Comment (bfolie) [reply src/citrine/_serialization/properties.py:235]: It just strikes me as a strange way to specify. If you want the cache size to be about a million, why not just say `maxsize=1e6`?

File: tests/_serialization/test_object_serialization.py
Comment (kroenlein) [reply tests/_serialization/test_object_serialization.py:81]: Linters complain when a object getter is called in null context.  This tells the linter you know what you are asking for without creating a new symbol.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py:235]: A bias towards base 2, but there's no compelling reason.

File: tests/_serialization/test_object_serialization.py
Comment (bfolie) [reply tests/_serialization/test_object_serialization.py:81]: Sorry, I should have been more precise. Why does calling the getting throw an Attribute Error?

Comment (kroenlein) [reply tests/_serialization/test_object_serialization.py:81]: Sorry, yes, very different question.  In a perfect world, this would be a compile time error because what's getting caught is that you've asked to override a method that doesn't appear in a parent class.  However, when the Property is defined, it doesn't actually know its parentage, and even if it did, it wouldn't know to which class variable you are going to assign it.  The easiest solution then is to throw an error at runtime when you can actually introspect enough to resolve an MRO.

File: src/citrine/_serialization/properties.py
Comment (kroenlein) [reply src/citrine/_serialization/properties.py]: Benchmarking showed saturation around 128, so I changed both caches down to 1024.  If you have strong opinions, I can change that to `2**10` or `10**3` (both are which are int)

**Comment by bfolie** [comment]
Let me know once you've run the e2es and then I'll approve

**Comment by kroenlein** [comment]
Dev e2es:
https://jenkins.corp.citrine.io/job/nextgen-ITs/37703

Stage e2es:
https://jenkins.corp.citrine.io/job/nextgen-ITs/37704

**Review by bfolie — APPROVED** [review]
Looks good, but @anoto-moniz should weigh in as well

File: src/citrine/resources/condition_template.py
Comment (anoto-moniz) [inline src/citrine/resources/condition_template.py:51]: This section is a bit odd, though it predates this commit. I see why we invoke `__init__` directly on separate classes, but why not use super to ensure we respect the parent's MRO and anything in between?

`super(AttributeTemplate, self).__init__()`

Comment (kroenlein) [reply src/citrine/resources/condition_template.py:51]: Dunno if I've ever seen that pattern preferred, but I don't see any harm.

File: tests/_serialization/test_simple_properties.py
Comment (anoto-moniz) [inline tests/_serialization/test_simple_properties.py:243]: This appears to be an interface break. If so, we need to either force it to a `ValueError` for the time being and make the switch come 3.0, or make sure we very clearly and proactively (read: not just release notes) communicate when it happens, so people can update their scripts if need be.

File: tests/resources/test_object_setters.py
Comment (anoto-moniz) [inline tests/resources/test_object_setters.py:56]: Same thing here as with LOE above.

File: src/citrine/_serialization/properties.py
Comment (anoto-moniz) [inline src/citrine/_serialization/properties.py:30]: ???

File: tests/resources/test_object_setters.py
Comment (kroenlein) [reply tests/resources/test_object_setters.py:56]: The problem here is that gemd-python (correctly) throws TypeErrors when passing arguments of the wrong type, and _serialization throws ValueErrors.  This change is reflective of the new MRO.  

It's even a little grosser than that, since the error message thrown is actually misleading.
```
ValueError: <class 'citrine.resources.measurement_run.MeasurementRun'> <Measurement run 'A measurement'> is not one of valid types: (<class 'dict'>, <class 'NoneType'>) for MaterialRun:process
```

File: tests/_serialization/test_simple_properties.py
Comment (kroenlein) [reply tests/_serialization/test_simple_properties.py:243]: Ironically, solvable with the try-except block that @bfolie asked me to replace with the signature test, albeit in this case the exception in thrown by a dependency.  Thinking about how to address this, given the harder issue below, but it's worth noting that this case should really be a TypeError because missing __init__ arguments are TypeErrors in the Python core libraries.

Comment (anoto-moniz) [reply tests/_serialization/test_simple_properties.py:243]: It makes sense to me this should be a `TypeError`, but that means a user's `try-except` won't catch it any more. So as annoying as it is, I do think we should catch it, and make a note to change it at the 3.0 release.

File: tests/resources/test_object_setters.py
Comment (anoto-moniz) [reply tests/resources/test_object_setters.py]: Ugh, I've seen that error message and it's awful.

Sounds like the underlying issue is that incongruence, and it will likely be more complicated to preserve as is now. Is `deserialize` an operation users will often call, or is it more an internally used method?

We should move `_serialization` to `TypeError` when we go to 3.0.

**Pull request merged**
